File: apps/games-flutter/lib/utils/word_search_generator.dart
import 'dart:math';
import '../models/word.dart';

class WordSearchGenerator {
  static const List<Direction> _directions = [
    Direction(0, 1), // horizontal →
    Direction(1, 0), // vertical ↓
    Direction(1, 1), // diagonal ↘
    Direction(1, -1), // diagonal ↙
  ];

  static WordSearchGrid generate(List<String> words, {int size = 12}) {
    final grid = List.generate(size, (_) => List<String>.filled(size, ''));
    final placedWords = <Word>[];

    // Normaliser et filtrer les mots
    final normalizedWords = words
        .map((w) => _normalizeWord(w.toUpperCase()))
        .where((w) => w.length >= 3 && w.length <= size)
        .take(10)
        .toList();

    // Placer les mots
    for (final word in normalizedWords) {
      final placed = _placeWord(grid, word, size);
      if (placed != null) {
        placedWords.add(placed);
      }
    }

    // Remplir les cases vides
    for (int row = 0; row < size; row++) {
      for (int col = 0; col < size; col++) {
        if (grid[row][col].isEmpty) {
          grid[row][col] = _getRandomLetter();
        }
      }
    }

    return WordSearchGrid(
      grid: grid,
      words: placedWords,
      size: size,
    );
  }

  static Word? _placeWord(List<List<String>> grid, String word, int size) {
    final random = Random();
    const maxAttempts = 100;

    for (int attempt = 0; attempt < maxAttempts; attempt++) {
      final direction = _directions[random.nextInt(_directions.length)];
      final startRow = random.nextInt(size);
      final startCol = random.nextInt(size);

      if (_canPlaceWord(grid, word, startRow, startCol, direction, size)) {
        final cells = <CellPosition>[];

        for (int i = 0; i < word.length; i++) {
          final row = startRow + i * direction.dx;
          final col = startCol + i * direction.dy;
          grid[row][col] = word[i];
          cells.add(CellPosition(row, col));
        }

        return Word(text: word, cells: cells);
      }
    }

    return null;
  }

  static bool _canPlaceWord(
    List<List<String>> grid,
    String word,
    int startRow,
    int startCol,
    Direction direction,
    int size,
  ) {
    for (int i = 0; i < word.length; i++) {
      final row = startRow + i * direction.dx;
      final col = startCol + i * direction.dy;

      // Vérifier les limites
      if (row < 0 || row >= size || col < 0 || col >= size) {
        return false;
      }

      // Vérifier si la case est vide ou contient la même lettre
      if (grid[row][col].isNotEmpty && grid[row][col] != word[i]) {
        return false;
      }
    }

    return true;
  }

  static String _normalizeWord(String word) {
    return word.replaceAll(RegExp(r'[^A-ZÀÁÂÃÄÅÈÉÊËÌÍÎÏÒÓÔÕÖÙÚÛÜÇÑòóôõöùúûüèéêëàáâãäåìíîïçñ]'), '').toUpperCase();
  }

  static String _getRandomLetter() {
    const letters = 'ABCDEFGHIJKLMNOPQRSTUVWXYZÉÈÊÀÔÙ';
    return letters[Random().nextInt(letters.length)];
  }
}

class Direction {
  final int dx;
  final int dy;

  const Direction(this.dx, this.dy);
}

File: apps/web-vitrine/lib/games/word-search-generator.ts
export interface Word {
  word: string;
  found: boolean;
  cells: { row: number; col: number }[];
}

export interface WordSearchGrid {
  grid: string[][];
  words: Word[];
  gridSize: number;
}

const DIRECTIONS = [
  { dx: 0, dy: 1 },   // horizontal →
  { dx: 1, dy: 0 },   // vertical ↓
  { dx: 1, dy: 1 },   // diagonal ↘
  { dx: 1, dy: -1 },  // diagonal ↙
];

export function generateWordSearch(words: string[], gridSize: number = 12): WordSearchGrid {
  const grid: string[][] = Array(gridSize).fill(null).map(() => Array(gridSize).fill(''));
  const placedWords: Word[] = [];

  // Normaliser et filtrer les mots
  const normalizedWords = words
    .map(w => normalizeCreoleWord(w.toUpperCase()))
    .filter(w => w.length >= 3 && w.length <= gridSize)
    .slice(0, 10); // Limiter à 10 mots max

  // Placer les mots
  for (const word of normalizedWords) {
    const placed = placeWord(grid, word, gridSize);
    if (placed) {
      placedWords.push(placed);
    }
  }

  // Remplir les cases vides avec des lettres aléatoires
  for (let row = 0; row < gridSize; row++) {
    for (let col = 0; col < gridSize; col++) {
      if (!grid[row][col]) {
        grid[row][col] = getRandomLetter();
      }
    }
  }

  return {
    grid,
    words: placedWords,
    gridSize,
  };
}

function placeWord(
  grid: string[][],
  word: string,
  gridSize: number
): Word | null {
  const attempts = 100;

  for (let attempt = 0; attempt < attempts; attempt++) {
    const direction = DIRECTIONS[Math.floor(Math.random() * DIRECTIONS.length)];
    const startRow = Math.floor(Math.random() * gridSize);
    const startCol = Math.floor(Math.random() * gridSize);

    if (canPlaceWord(grid, word, startRow, startCol, direction, gridSize)) {
      const cells: { row: number; col: number }[] = [];

      for (let i = 0; i < word.length; i++) {
        const row = startRow + i * direction.dx;
        const col = startCol + i * direction.dy;
        grid[row][col] = word[i];
        cells.push({ row, col });
      }

      return {
        word,
        found: false,
        cells,
      };
    }
  }

  return null;
}

function canPlaceWord(
  grid: string[][],
  word: string,
  startRow: number,
  startCol: number,
  direction: { dx: number; dy: number },
  gridSize: number
): boolean {
  for (let i = 0; i < word.length; i++) {
    const row = startRow + i * direction.dx;
    const col = startCol + i * direction.dy;

    // Vérifier les limites
    if (row < 0 || row >= gridSize || col < 0 || col >= gridSize) {
      return false;
    }

    // Vérifier si la case est vide ou contient la même lettre
    if (grid[row][col] && grid[row][col] !== word[i]) {
      return false;
    }
  }

  return true;
}

function normalizeCreoleWord(word: string): string {
  // Supprimer les caractères spéciaux et garder seulement les lettres créoles
  return word
    .replace(/[^A-ZÀÁÂÃÄÅÈÉÊËÌÍÎÏÒÓÔÕÖÙÚÛÜÇÑòóôõöùúûüèéêëàáâãäåìíîïçñ]/g, '')
    .toUpperCase();
}

function getRandomLetter(): string {
  // Lettres courantes en créole martiniquais
  const letters = 'ABCDEFGHIJKLMNOPQRSTUVWXYZÉÈÊÀÔÙ';
  return letters[Math.floor(Math.random() * letters.length)];
}

// Fonction pour extraire des mots aléatoires du dictionnaire
export function getRandomWordsFromDict(dictData: any[], count: number = 10): string[] {
  const shuffled = [...dictData].sort(() => 0.5 - Math.random());
  return shuffled.slice(0, count).map(item => item.mot);
}
